pennylane/qinfo/transforms.py

```diff
@@ -396,3 +396,104 @@ def wrapper(*args, **kwargs):
         return _compute_cfim(p, j)
 
     return wrapper
+
+
+def fidelity(qnode0, qnode1, wires0, wires1):
+    r"""Compute the fidelity for two :class:`.QNode` returning a :func:`~.state` (a state can be a state vector
+    or a density matrix, depending on the device) acting on quantum systems with the same size. For two pure states, the
+    fidelity corresponds to the squared overlap. For a pure state and a mixed state, it corresponds to the squared
+    expectation of the mixed state in the pure state. Finally for two mixed states, it is defined by the last formula:
+
+    .. math::
+        \vspace \text{ (1)} F( \ket{\psi} , \ket{\phi}) = \left|\bra{\psi}\ket{\phi}\right|^2
+
+        \vspace \text{ (2)} F( \ket{\psi} , \sigma ) = \left|\bra{\psi} \sigma \ket{\psi}\right|^2
+
+        \vspace \text{ (3)} F( \rho , \sigma ) = -\text{Tr}( \sqrt{\sqrt{\rho} \sigma \sqrt{\rho}})^2
+
+    .. warning::
+        The second state is coerced to the type and dtype of the first state. The fidelity is returned in the type
+        of the interface of the first state.
+
+    Args:
+        state0 (QNode): A :class:`.QNode` returning a :func:`~.state`.
+        state1 (QNode): A :class:`.QNode` returning a :func:`~.state`.
+        wires0 (Sequence[int]): the wires of the first subsystem
+        wires1 (Sequence[int]): the wires of the second subsystem
+
+    Returns:
+        func: A function with the same arguments as the QNodes that returns the fidelities between the output states.
```

Comment:
```suggestion
        func: A function that returns the fidelity between the states outputted by the QNodes.
```

Reply:
> with the same arguments as the QNodes

I think this reads a bit odd and instead should be elaborated on in the docstring itself. Also: we should definitely have an example for when one QNode doesn't take arguments. For such a case, what would happen if we passed in an empty tuple?

I.e., do 
```py
@qml.qnode(dev)
def circuit_rx():
    return qml.state()

@qml.qnode(dev)
def circuit_ry(x):
    qml.RY(x, wires=0)
    return qml.state()

qml.qinfo.fidelity(circuit_rx, circuit_ry, wires0=[0], wires1=[0])((,), (0.2))
```
The `all_args` syntax is not visible to the users. If that's the only way to handle the case above, it should be documented more.

Reply:
Totally agree I will add it 👍